// xiome/x/features/auth/aspects/permissions/tools/prepare-stats-hub.js
import * as renraku from "renraku";
import * as dbmage from "dbmage";
import { day, month } from "../../../../../toolbox/goodtimes/times.js";
export function prepareStatsHub({ database }) {
    return async function getStatsHub(userId) {
        async function throwForbiddenUser(appId) {
            const row = await database.tables.apps.owners.readOne(dbmage.find({ appId }));
            if (row.userId.toString() !== userId.toString())
                throw new renraku.ApiError(403, "forbidden");
        }
        return {
            countUsers: async (appId) => {
                await throwForbiddenUser(appId);
                const accountsTable = database.tables.auth.users.accounts
                    .constrainForApp(appId);
                return accountsTable.count({ conditions: false });
            },
            countUsersActiveDaily: async (appId) => {
                await throwForbiddenUser(appId);
                const timeToStartCounting = Date.now() - day;
                const latestLoginsTable = database.tables.auth.users.latestLogins
                    .constrainForApp(appId);
                return latestLoginsTable.count({
                    conditions: dbmage.and({ greater: { time: timeToStartCounting } }),
                });
            },
            countUsersActiveMonthly: async (appId) => {
                await throwForbiddenUser(appId);
                const timeToStartCounting = Date.now() - month;
                const latestLoginsTable = database.tables.auth.users.latestLogins
                    .constrainForApp(appId);
                return latestLoginsTable.count({
                    conditions: dbmage.and({ greater: { time: timeToStartCounting } }),
                });
            },
        };
    };
}
