// xiome/x/toolbox/popups/setup-manual-popup.js
import { PopupFlag } from "./types/popup-flag.js";
import { validateRequest } from "./validation/validate-request.js";
export function setupManualPopup({ namespace, allowedOrigins, }) {
    const opener = window.opener;
    if (!opener)
        throw new Error(`popup "${namespace}" cannot act standalone`);
    function sendResponse(targetOrigin, message) {
        opener.postMessage(message, targetOrigin);
    }
    return {
        respondError(targetOrigin, error) {
            sendResponse(targetOrigin, {
                error,
                namespace,
                flag: PopupFlag.ErrorResponse,
            });
        },
        respondPayload(targetOrigin, payload) {
            sendResponse(targetOrigin, {
                namespace,
                payload: payload,
                flag: PopupFlag.PayloadResponse,
            });
        },
        sendReadyAndListenForGo(action) {
            window.addEventListener("message", async function goListener(event) {
                if (!validateRequest({
                    namespace,
                    event,
                    allowedOrigins,
                }))
                    return null;
                window.removeEventListener("message", goListener);
                action(event.origin, event.data.parameters);
            });
            sendResponse("*", {
                namespace,
                flag: PopupFlag.ReadyResponse,
            });
        },
    };
}
