// xiome/x/features/store/popups/make-stripe-popup-spec.js
import { encodeQuerystring } from "./utils/encode-querystring.js";
export var makeStripePopupSpec;
(function (makeStripePopupSpec) {
    function connect({ popupReturnUrl, generateId }) {
        const popupId = generateId().string;
        return {
            popupId,
            return_url: `${popupReturnUrl}?${encodeQuerystring({
                popupId,
                status: "return",
            })}`,
            refresh_url: `${popupReturnUrl}?${encodeQuerystring({
                popupId,
                status: "refresh",
            })}`,
        };
    }
    makeStripePopupSpec.connect = connect;
    function login({ generateId }) {
        return { popupId: generateId().string };
    }
    makeStripePopupSpec.login = login;
    function openCustomerPortal({ generateId, popupReturnUrl }) {
        const popupId = generateId().string;
        return {
            popupId,
            return_url: `${popupReturnUrl}?${encodeQuerystring({
                popupId,
                status: "return",
            })}`
        };
    }
    makeStripePopupSpec.openCustomerPortal = openCustomerPortal;
    function checkout({ popupReturnUrl, generateId }) {
        const popupId = generateId().string;
        return {
            popupId,
            success_url: `${popupReturnUrl}?${encodeQuerystring({
                popupId,
                status: "success",
            })}`,
            cancel_url: `${popupReturnUrl}?${encodeQuerystring({
                popupId,
                status: "cancel",
            })}`,
        };
    }
    makeStripePopupSpec.checkout = checkout;
})(makeStripePopupSpec || (makeStripePopupSpec = {}));
