// xiome/x/features/videos/models/parts/content-model.js
import { snapstate } from "@chasemoskal/snapstate";
import { ops } from "../../../../framework/ops.js";
import { videoPrivileges } from "../../api/video-privileges.js";
export function makeContentModel({ contentService, }) {
    const state = snapstate({
        accessOp: ops.none(),
        catalogOp: ops.none(),
        viewsOp: ops.none(),
        privilegesOp: ops.none(),
        showsOp: ops.none(),
    });
    const activeShowLabels = new Set();
    const currentlyLoadingShows = new Set();
    function getAccess() {
        return ops.value(state.readable.accessOp);
    }
    let initialized = false;
    async function loadModerationData() {
        initialized = true;
        const access = getAccess();
        const isModerator = access && access.permit
            .privileges.includes(videoPrivileges["moderate videos"]);
        if (isModerator) {
            await ops.operation({
                promise: contentService.fetchModerationData(),
                setOp: op => {
                    var _a, _b, _c;
                    const data = ops.value(op);
                    state.writable.catalogOp = ops.replaceValue(op, (_a = data === null || data === void 0 ? void 0 : data.catalog) !== null && _a !== void 0 ? _a : []);
                    state.writable.viewsOp = ops.replaceValue(op, (_b = data === null || data === void 0 ? void 0 : data.views) !== null && _b !== void 0 ? _b : []);
                    state.writable.privilegesOp = ops.replaceValue(op, (_c = data === null || data === void 0 ? void 0 : data.privileges) !== null && _c !== void 0 ? _c : []);
                },
            });
        }
    }
    async function loadShow(label) {
        var _a;
        initialized = true;
        activeShowLabels.add(label);
        if (!getAccess() || currentlyLoadingShows.has(label))
            return undefined;
        currentlyLoadingShows.add(label);
        const oldShows = (_a = ops.value(state.readable.showsOp)) !== null && _a !== void 0 ? _a : [];
        let updatedShow;
        await ops.operation({
            setOp: op => state.writable.showsOp = op,
            promise: contentService.getShows({ labels: [label] })
                .then(shows => shows[0])
                .then(show => updatedShow = show)
                .then(show => [
                ...oldShows.filter(s => s.label !== label),
                show,
                // ...(show ? [show] : [{label, details: undefined}]),
            ]),
        });
        currentlyLoadingShows.delete(label);
        return updatedShow;
    }
    async function refreshShows() {
        initialized = true;
        if (!getAccess())
            return undefined;
        const labels = Array.from(activeShowLabels)
            .filter(label => !currentlyLoadingShows.has(label));
        if (labels.length) {
            for (const label of labels)
                currentlyLoadingShows.add(label);
            await ops.operation({
                setOp: op => state.writable.showsOp = op,
                promise: contentService.getShows({ labels })
                    .then(shows => shows.filter(s => !!s))
            });
            for (const label of labels)
                currentlyLoadingShows.delete(label);
        }
    }
    return {
        state: state.readable,
        subscribe: state.subscribe,
        async updateAccessOp(op) {
            state.writable.accessOp = op;
            state.writable.catalogOp = ops.none();
            state.writable.viewsOp = ops.none();
            if (initialized) {
                await loadModerationData();
                await refreshShows();
            }
        },
        async onVideoHostingUpdate() {
            if (initialized) {
                await loadModerationData();
                await refreshShows();
            }
        },
        async initializeForModerationData() {
            if (!initialized) {
                await loadModerationData();
            }
        },
        async initializeForVideo(label) {
            if (!initialized)
                await loadModerationData();
            await loadShow(label);
        },
        get allowance() {
            const access = ops.value(state.readable.accessOp);
            const can = (p) => access
                ? access.permit.privileges.includes(videoPrivileges[p])
                : false;
            return {
                canModerateVideos: can("moderate videos"),
                canViewAllVideos: can("moderate videos") || can("view all videos"),
            };
        },
        get catalog() {
            var _a;
            return (_a = ops.value(state.readable.catalogOp)) !== null && _a !== void 0 ? _a : [];
        },
        get views() {
            var _a;
            return (_a = ops.value(state.readable.viewsOp)) !== null && _a !== void 0 ? _a : [];
        },
        get privileges() {
            var _a;
            return (_a = ops.value(state.readable.privilegesOp)) !== null && _a !== void 0 ? _a : [];
        },
        get shows() {
            var _a;
            return (_a = ops.value(state.readable.showsOp)) !== null && _a !== void 0 ? _a : [];
        },
        getView(label) {
            var _a;
            return ((_a = ops.value(state.readable.viewsOp)) !== null && _a !== void 0 ? _a : [])
                .find(view => view.label === label);
        },
        getPrivilege(id) {
            var _a;
            return ((_a = ops.value(state.readable.privilegesOp)) !== null && _a !== void 0 ? _a : [])
                .find(p => p.privilegeId === id);
        },
        getShow(label) {
            var _a;
            return ((_a = ops.value(state.readable.showsOp)) !== null && _a !== void 0 ? _a : [])
                .find(show => show.label === label);
        },
        async setView(view) {
            var _a;
            const oldViews = (_a = ops.value(state.readable.viewsOp)) !== null && _a !== void 0 ? _a : [];
            await ops.operation({
                setOp: op => state.writable.viewsOp = op,
                promise: contentService
                    .writeView(view)
                    .then(() => [
                    ...oldViews.filter(v => v.label !== view.label),
                    view,
                ]),
            });
            await loadShow(view.label);
        },
        async deleteView(label) {
            var _a, _b;
            const oldViews = (_a = ops.value(state.readable.viewsOp)) !== null && _a !== void 0 ? _a : [];
            await ops.operation({
                setOp: op => state.writable.viewsOp = op,
                promise: contentService
                    .deleteView({ label })
                    .then(() => oldViews.filter(v => v.label !== label)),
            });
            const oldShows = (_b = ops.value(state.readable.showsOp)) !== null && _b !== void 0 ? _b : [];
            state.writable.showsOp = ops.replaceValue(state.readable.showsOp, oldShows.filter(s => s.label !== label));
        },
    };
}
