// xiome/x/assembly/backend/database/database-shapes.js
export const databaseShapeUnisolated = {
    apps: {
        registrations: true,
        owners: true,
    },
};
export const databaseShapeRequiresAppIsolation = {
    auth: {
        users: {
            accounts: true,
            emails: true,
            latestLogins: true,
            profiles: true,
        },
        permissions: {
            privilege: true,
            role: true,
            roleHasPrivilege: true,
            userHasRole: true,
        },
    },
    example: {
        examplePosts: true,
    },
    questions: {
        likes: true,
        reports: true,
        answerPosts: true,
        questionPosts: true,
    },
    store: {
        customers: true,
        connect: {
            accounts: true,
            active: true,
        },
        subscriptions: {
            plans: true,
            tiers: true,
        },
    },
    videos: {
        dacastAccountLinks: true,
        viewDacast: true,
        viewPrivileges: true,
    },
    notes: {
        notes: true,
        questionDetails: true,
    },
};
export const databaseShape = {
    ...databaseShapeRequiresAppIsolation,
    ...databaseShapeUnisolated,
};
