// xiome/x/features/auth/aspects/users/models/access-model.js
import { pub } from "../../../../../toolbox/pub.js";
import { clone } from "../../../../../toolbox/clone.js";
import { ops } from "../../../../../framework/ops.js";
import { AccessLoginExpiredError } from "./errors/access-errors.js";
import { isTokenValid } from "../../../utils/tokens/is-token-valid.js";
export function makeAccessModel({ authMediator, loginService }) {
    let accessOp = ops.none();
    const accessOpEvent = pub();
    async function setAccessOp(op) {
        accessOp = op;
        await accessOpEvent.publish(op);
    }
    authMediator.subscribeToAccessChange(async (access) => setAccessOp(ops.ready(access)));
    async function accessOperation(promise) {
        return ops.operation({
            promise,
            setOp: setAccessOp,
        });
    }
    const loginFacilities = {
        async useExistingLogin() {
            await accessOperation(authMediator.initialize());
        },
        async sendLoginLink(email) {
            return loginService.sendLoginLink({ email });
        },
        async login(loginToken) {
            try {
                if (isTokenValid(loginToken))
                    await accessOperation(loginService
                        .authenticateViaLoginToken({ loginToken })
                        .then(tokens => authMediator.login(tokens)));
                else
                    throw new AccessLoginExpiredError();
            }
            catch (error) {
                console.error(error);
                await setAccessOp(ops.none());
                await accessOperation(authMediator.initialize());
                throw error;
            }
        },
        async logout() {
            await accessOperation(authMediator.logout());
            await ops.operation({
                promise: authMediator.logout(),
                setOp: setAccessOp,
            });
        },
        async reauthorize() {
            await accessOperation(authMediator.reauthorize());
        },
    };
    return {
        ...loginFacilities,
        subscribe: accessOpEvent.subscribe,
        getAccessOp() {
            return clone(accessOp);
        },
        getAccess() {
            return ops.value(clone(accessOp));
        },
        getValidAccess() {
            return authMediator.getValidAccess();
        },
    };
}
