// xiome/x/features/store/testing/utils/common-setups.js
import { storeTestSetup } from "./store-test-setup.js";
export async function connectedStore() {
    const api = await storeTestSetup()
        .then(x => x.api());
    const merchant = await api
        .client(api.roles.merchant)
        .then(x => x.browserTab());
    await merchant
        .store
        .connect
        .stripeAccountOnboarding();
    return { api, merchant };
}
export async function storeWithSubscriptionPlans() {
    const { api } = await connectedStore();
    const clerk = await api
        .client(api.roles.clerk)
        .then(x => x.browserTab());
    const { planId } = await clerk
        .store
        .subscriptions
        .addPlan({
        planLabel: "membership",
        tier: {
            label: "benevolent donor",
            pricing: {
                currency: "usd",
                interval: "month",
                price: 500,
            },
        },
    });
    await clerk
        .store
        .subscriptions
        .addTier({
        planId,
        label: "deluxe",
        pricing: {
            currency: "usd",
            interval: "month",
            price: 10000,
        },
    });
    await clerk
        .store
        .subscriptions
        .addPlan({
        planLabel: "bees",
        tier: {
            label: "worker bee",
            pricing: {
                currency: "usd",
                interval: "month",
                price: 400,
            },
        },
    });
    function getMySubscription(store, tierId) {
        return store.get
            .subscriptions
            .mySubscriptionDetails
            .find(subscription => subscription.tierId === tierId);
    }
    return { api, clerk, getMySubscription };
}
