// xiome/x/assembly/frontend/modal/setup-modal-system.js
import { trapFocus } from "./utils/trap-focus.js";
import { preparePrompt } from "./prefabs/prompt.js";
import { prepareConfirm } from "./prefabs/confirm.js";
import { html } from "../../../framework/component.js";
import { prepareModalSystemRendering } from "./utils/prepare-modal-system-rendering.js";
import { prepareAlert } from "./prefabs/alert.js";
import { prepareConfirmAction } from "./prefabs/confirm-action.js";
export function setupModalSystem() {
    let count = 0;
    const modals = new Map();
    const listModals = () => Array.from(modals.values());
    const { rerender, element } = prepareModalSystemRendering(listModals);
    function popup({ focusNthElement, renderContent, onBlanketClick }) {
        var _a, _b, _c;
        const id = count++;
        const controls = {
            rerender,
            close: () => {
                modals.delete(id);
                rerender();
            },
        };
        const top = (_c = (_b = (_a = window.pageYOffset) !== null && _a !== void 0 ? _a : document.documentElement.scrollTop) !== null && _b !== void 0 ? _b : document.body.scrollTop) !== null && _c !== void 0 ? _c : 0;
        const handleBlanketClick = () => onBlanketClick(controls);
        modals.set(id, {
            render() {
                return html `
					<div data-modal="${id}">
						<div data-blanket @click=${handleBlanketClick}></div>
						<div data-plate style="top: ${top}px">
							${renderContent(controls)}
						</div>
					</div>
				`;
            },
        });
        rerender();
        const modal = element.querySelector(`[data-modal="${id}"]`);
        trapFocus(modal, focusNthElement);
        return { controls, modal };
    }
    const confirm = prepareConfirm(popup);
    const confirmAction = prepareConfirmAction(popup);
    return {
        modalsElement: element,
        modals: {
            popup,
            confirm,
            confirmAction,
            alert: prepareAlert(confirm),
            prompt: preparePrompt(popup),
        },
    };
}
