// renraku/s/types.ts

import type {IncomingHttpHeaders} from "http"

export interface Logger {
	log(...data: any[]): void
	warn(...data: any[]): void
	error(...data: any[]): void
}

export interface Methods {
	[key: string]: (...args: any[]) => Promise<any>
}

export interface HttpHeaders extends IncomingHttpHeaders {}

export interface Policy<xMeta, xAuth> {
	(meta: xMeta, headers?: HttpHeaders): Promise<xAuth>
}

export interface Expose<xAuth, xMethods extends Methods> {
	(auth: xAuth): xMethods
}

export const is_service = Symbol("is_service")

export interface Service<xMeta, xAuth, xMethods extends Methods> {
	[is_service]: symbol
	policy: Policy<xMeta, xAuth>
	expose: Expose<xAuth, xMethods>
}

export interface Api {
	[key: string]: Api | Service<any, any, Methods>
}

export type MetaMap<xApi extends Api> = {
	[P in keyof xApi]: xApi[P] extends Service<infer xMeta, any, Methods>
		? () => Promise<xMeta>
		: xApi[P] extends Api
			? MetaMap<xApi[P]>
			: never
}

export type AuthMap<xApi extends Api> = {
	[P in keyof xApi]: xApi[P] extends Service<any, infer xAuth, Methods>
		? () => Promise<xAuth>
		: xApi[P] extends Api
			? AuthMap<xApi[P]>
			: never
}

export type ApiRemote<xApi extends Api> = {
	[P in keyof xApi]: xApi[P] extends Service<any, any, infer xMethods>
		? xMethods
		: xApi[P] extends Api
			? ApiRemote<xApi[P]>
			: never
}

export type Remote<X extends Service<any, any, any> | Api> =
	X extends Service<any, any, infer xMethods>
		? xMethods
		: X extends Api
			? ApiRemote<X>
			: never

export interface Request {
	meta: any
	method: string
	params: any[]
	headers?: HttpHeaders
}

export interface Response {
	result: any
}

export interface Requester {
	(request: Request): Promise<any>
}

export interface JsonRpcRequest {
	jsonrpc: "2.0"
	id: number
	method: string
	params: any[]
}

export interface JsonRpcRequestWithMeta extends JsonRpcRequest {
	meta: any
}

export interface JsonRpcResponseCommon {
	jsonrpc: "2.0"
	id: number
}

export interface JsonRpcSuccessResponse extends JsonRpcResponseCommon {
	result: any
}

export interface JsonRpcErrorResponse extends JsonRpcResponseCommon {
	error: {
		code: number
		message: string
		data?: any
	}
}

export type JsonRpcResponse = JsonRpcSuccessResponse | JsonRpcErrorResponse

export interface Servelet {
	(request: Request): Promise<any>
}

export interface Requester {
	(request: Request): Promise<any>
}

export interface ConnectionControls {
	ping(): void
	close(): void
}

export interface MockLatency {
	min: number
	max: number
}

export interface MockOptions {
	getMockLatency?: () => undefined | MockLatency
}

export interface Spike {
	(methodName: string, func: (...params: any[]) => Promise<any>, ...params: any): Promise<any>
}

export interface ServiceOptions {
	spike?: Spike
}
