// dbmage/x/drivers/mongo/conversions.js
import { Binary } from "mongodb";
import { Id } from "../../id.js";
import { objectMap } from "../../tools/object-map.js";
// strip away the mongo database id's -- we don't use 'em
function skimMongoId(row) {
    if (row) {
        const { _id: noop, ...skimmed } = row;
        return skimmed;
    }
    return undefined;
}
export function valueUp(value, key) {
    return value instanceof Id
        ? new Binary(Buffer.from(value.binary))
        : value;
}
export function valueDown(value, key) {
    return value instanceof Binary
        ? new Id(value.buffer)
        : value;
}
// process a row before it's sent to mongo
// - transform any id to mongo binary type
export function up(row) {
    return objectMap(row, valueUp);
}
// process a row retrieved from mongo
// - transform any binary types into id
export function down(data) {
    return (data && typeof data === "object")
        ? objectMap(skimMongoId(data), valueDown)
        : data;
}
export function ups(rows) {
    return rows.map(up);
}
export function downs(rows) {
    return rows.map(down);
}
