// xiome/x/features/store/isomorphic/concepts.js
export var StripeConnectStatus;
(function (StripeConnectStatus) {
    StripeConnectStatus[StripeConnectStatus["Unlinked"] = 0] = "Unlinked";
    StripeConnectStatus[StripeConnectStatus["Incomplete"] = 1] = "Incomplete";
    StripeConnectStatus[StripeConnectStatus["Paused"] = 2] = "Paused";
    StripeConnectStatus[StripeConnectStatus["Ready"] = 3] = "Ready";
})(StripeConnectStatus || (StripeConnectStatus = {}));
export var SubscriptionStatus;
(function (SubscriptionStatus) {
    SubscriptionStatus[SubscriptionStatus["Unsubscribed"] = 0] = "Unsubscribed";
    SubscriptionStatus[SubscriptionStatus["Active"] = 1] = "Active";
    SubscriptionStatus[SubscriptionStatus["Unpaid"] = 2] = "Unpaid";
    SubscriptionStatus[SubscriptionStatus["Cancelled"] = 3] = "Cancelled";
})(SubscriptionStatus || (SubscriptionStatus = {}));
export var PurchaseScenario;
(function (PurchaseScenario) {
    PurchaseScenario[PurchaseScenario["Update"] = 0] = "Update";
    PurchaseScenario[PurchaseScenario["UsePaymentMethod"] = 1] = "UsePaymentMethod";
    PurchaseScenario[PurchaseScenario["CheckoutPopup"] = 2] = "CheckoutPopup";
})(PurchaseScenario || (PurchaseScenario = {}));
