// xiome/x/features/store/isomorphic/validators.js
import { validateId } from "../../../common/validators/validate-id.js";
import { boolean, branch, is, maxLength, min, minLength, number, schema, string, validator } from "../../../toolbox/darkvalley.js";
export const validatePriceNumber = validator(number(), min(0.01));
export const validateLabel = validator(string(), minLength(1), maxLength(32));
export const validateCurrency = validator(string(), branch(is("usd")));
export const validateInterval = validator(string(), branch(is("month"), is("year")));
export const validateSubscriptionPricingDraft = schema({
    price: validatePriceNumber,
    currency: validateCurrency,
    interval: validateInterval,
});
export const validateBoolean = validator(boolean());
export const validatePricing = schema({
    currency: validateCurrency,
    interval: validateInterval,
    price: validatePriceNumber,
});
export const validateNewTierDraft = schema({
    label: validateLabel,
    pricing: validatePricing,
});
export const validateNewPlanDraft = schema({
    planLabel: validateLabel,
    tier: validateNewTierDraft,
});
export const validateEditPlanDraft = schema({
    planId: validateId,
    label: validateLabel,
    archived: validateBoolean,
});
export const validateEditTierDraft = schema({
    tierId: validateId,
    label: validateLabel,
    active: validateBoolean,
    pricing: validatePricing,
});
