// xiome/x/framework/api/unconstrained-table.js
import { objectMap } from "../../toolbox/object-map.js";
import * as dbmage from "dbmage";
import { appConstraintKey } from "../../assembly/backend/types/database.js";
export class UnconstrainedTable {
    static wrapTables(tables) {
        function recurse(t) {
            return objectMap(t, value => dbmage.isTable(value)
                ? new UnconstrainedTable(value)
                : recurse(value));
        }
        return recurse(tables);
    }
    static unwrapTables(unconstrainedTables) {
        function recurse(tables) {
            return objectMap(tables, value => value instanceof UnconstrainedTable
                ? value.unconstrained
                : recurse(value));
        }
        return recurse(unconstrainedTables);
    }
    static wrapDatabase(database) {
        return dbmage.subsection(database, tables => UnconstrainedTable.wrapTables(tables));
    }
    static constrainTablesForApp({ appId, unconstrainedTables, }) {
        function recurse(tables) {
            return objectMap(tables, value => value instanceof UnconstrainedTable
                ? value.constrainForApp(appId)
                : recurse(value));
        }
        return recurse(unconstrainedTables);
    }
    static constrainDatabaseForApp({ appId, database, }) {
        function recurse(tables) {
            return objectMap(tables, value => value instanceof UnconstrainedTable
                ? value.constrainForApp(appId)
                : dbmage.isTable(value)
                    ? value
                    : recurse(value));
        }
        return {
            tables: recurse(database.tables),
            transaction: (async (action) => database.transaction(async ({ tables, abort }) => action({
                tables: recurse(tables),
                abort,
            })))
        };
    }
    constructor(table) {
        this.unconstrained = table;
    }
    constrainForApp(appId) {
        return dbmage.constrain({
            table: this.unconstrained,
            constraint: { [appConstraintKey]: appId },
        });
    }
}
