// xiome/x/features/xio-components/price-display/xio-price-display.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _XioPriceDisplay_instances, _XioPriceDisplay_leftPadCents, _XioPriceDisplay_getPriceDetails, _XioPriceDisplay_renderWhenSuperscriptIsDisabled, _XioPriceDisplay_renderWhenSupercriptIsEnabled;
import styles from "./xio-price-display.css.js";
import { Component, html, property, mixinStyles } from "../../../framework/component.js";
let XioPriceDisplay = class XioPriceDisplay extends Component {
    constructor() {
        super(...arguments);
        _XioPriceDisplay_instances.add(this);
        this.symbol = "$";
        this.currency = "USD";
        this.value = "9.50";
        _XioPriceDisplay_renderWhenSuperscriptIsDisabled.set(this, () => {
            const { cents, dollars } = __classPrivateFieldGet(this, _XioPriceDisplay_instances, "m", _XioPriceDisplay_getPriceDetails).call(this);
            return html `
			<strong>
				${`${dollars}.${cents}`}
			</strong>
		`;
        });
        _XioPriceDisplay_renderWhenSupercriptIsEnabled.set(this, () => {
            const { cents, dollars } = __classPrivateFieldGet(this, _XioPriceDisplay_instances, "m", _XioPriceDisplay_getPriceDetails).call(this);
            return html `
			<strong>${dollars}</strong>
			<span class="superscript">${cents}</span>
		`;
        });
    }
    render() {
        const enableSuperscript = this["unit-superscript"];
        const enableVerticalCurrency = this["vertical-currency"];
        const { symbol, currency } = this;
        return html `
			<div class="card">
				<div class="display">
					<span class="symbol">${symbol}</span>
					${enableSuperscript
            ? __classPrivateFieldGet(this, _XioPriceDisplay_renderWhenSupercriptIsEnabled, "f").call(this)
            : __classPrivateFieldGet(this, _XioPriceDisplay_renderWhenSuperscriptIsDisabled, "f").call(this)}
					<span class=${`currency ${enableVerticalCurrency ? 'vertical' : ''}`}>
						${currency}
					</span>
				</div>
			</div>
		`;
    }
};
_XioPriceDisplay_renderWhenSuperscriptIsDisabled = new WeakMap();
_XioPriceDisplay_renderWhenSupercriptIsEnabled = new WeakMap();
_XioPriceDisplay_instances = new WeakSet();
_XioPriceDisplay_leftPadCents = function _XioPriceDisplay_leftPadCents(cents) {
    return cents.length > 1 ? cents : `0${cents}`;
};
_XioPriceDisplay_getPriceDetails = function _XioPriceDisplay_getPriceDetails() {
    const { value } = this;
    const numerical = parseFloat(value);
    const dollars = Math.floor(numerical);
    const cents = Math.round((numerical % 1.0) * 100);
    const strings = {
        dollars: dollars.toString(),
        cents: __classPrivateFieldGet(this, _XioPriceDisplay_instances, "m", _XioPriceDisplay_leftPadCents).call(this, cents.toString()),
    };
    return strings;
};
__decorate([
    property({ type: String, reflect: true })
], XioPriceDisplay.prototype, "symbol", void 0);
__decorate([
    property({ type: String, reflect: true })
], XioPriceDisplay.prototype, "currency", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], XioPriceDisplay.prototype, "unit-superscript", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], XioPriceDisplay.prototype, "vertical-currency", void 0);
__decorate([
    property({ type: String })
], XioPriceDisplay.prototype, "value", void 0);
XioPriceDisplay = __decorate([
    mixinStyles(styles)
], XioPriceDisplay);
export { XioPriceDisplay };
